streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KStream.scala

```diff
@@ -0,0 +1,203 @@
+import org.apache.kafka.streams.scala.FunctionConversions._
+
+import scala.collection.JavaConverters._
+
+/**
+ * Wraps the Java class KStream and delegates method calls to the underlying Java object.
+ */
+class KStream[K, V](val inner: KStreamJ[K, V]) {
+
+  def filter(predicate: (K, V) => Boolean): KStream[K, V] = {
+    inner.filter(predicate(_, _))
+  }
+
+  def filterNot(predicate: (K, V) => Boolean): KStream[K, V] = {
+    inner.filterNot(predicate(_, _))
+  }
+
+  def selectKey[KR](mapper: (K, V) => KR): KStream[KR, V] = {
+    inner.selectKey[KR]((k: K, v: V) => mapper(k, v))
+  }
+
+  def map[KR, VR](mapper: (K, V) => (KR, VR)): KStream[KR, VR] = {
+    val kvMapper = mapper.tupled andThen tuple2ToKeyValue
+    inner.map[KR, VR]((k, v) => kvMapper(k,v))
+  }
+
+  def mapValues[VR](mapper: V => VR): KStream[K, VR] = {
+    inner.mapValues[VR](mapper.asValueMapper)
+  }
+
+  def mapValues[VR](mapper: (K, V) => VR): KStream[K, VR] = {
+    inner.mapValues[VR](mapper.asValueMapperWithKey)
+  }
+
+  def flatMap[KR, VR](mapper: (K, V) => Iterable[(KR, VR)]): KStream[KR, VR] = {
+    val kvMapper = mapper.tupled andThen (iter => iter.map(tuple2ToKeyValue).asJava)
+    inner.flatMap[KR, VR]((k,v) => kvMapper(k , v))
+  }
+
+  def flatMapValues[VR](processor: V => Iterable[VR]): KStream[K, VR] = {
+    inner.flatMapValues[VR](processor.asValueMapper)
+  }
+
+  def flatMapValues[VR](processor: (K, V) => Iterable[VR]): KStream[K, VR] = {
+    inner.flatMapValues[VR](processor.asValueMapperWithKey)
+  }
+
+  def print(printed: Printed[K, V]): Unit = inner.print(printed)
+
+  def foreach(action: (K, V) => Unit): Unit = {
+    inner.foreach((k, v) => action(k, v))
+  }
+
+  def branch(predicates: ((K, V) => Boolean)*): Array[KStream[K, V]] = {
+    inner.branch(predicates.map(_.asPredicate): _*).map(kstream => wrapKStream(kstream))
+  }
+
+  def through(topic: String)(implicit produced: Produced[K, V]): KStream[K, V] =
+    inner.through(topic, produced)
+
+  def to(topic: String)(implicit produced: Produced[K, V]): Unit =
+    inner.to(topic, produced)
+
+  //scalastyle:off null
+  def transform[K1, V1](transformerSupplier: () => Transformer[K, V, (K1, V1)],
+    stateStoreNames: String*): KStream[K1, V1] = {
+
+    val transformerSupplierJ: TransformerSupplier[K, V, KeyValue[K1, V1]] = () => {
+      val transformerS: Transformer[K, V, (K1, V1)] = transformerSupplier()
+      new Transformer[K, V, KeyValue[K1, V1]] {
+        override def transform(key: K, value: V): KeyValue[K1, V1] = {
+          transformerS.transform(key, value) match {
+            case (k1, v1) => KeyValue.pair(k1, v1)
+            case _ => null
+          }
+        }
+
+        override def init(context: ProcessorContext): Unit = transformerS.init(context)
+
+        @deprecated ("Please use Punctuator functional interface at https://kafka.apache.org/10/javadoc/org/apache/kafka/streams/processor/Punctuator.html instead", "0.1.3") // scalastyle:ignore
```

Comment:
Nit: I am not a fan of adding links in JavaDocs, because links might break; better reference to the corresponding class as JavaDoc cross reference?

Also: I am wondering if we should remove this method from the wrapper in the first place? IMHO, it's not a good idea to add deprecated API in new code?

Reply:
Agreed.  I'll remove it.

Reply:
Here, the deprecation is on `punctuate`, which is part of the contract of `Transformer`. How do we remove this ? We can *only* remove this when `punctuate` is removed from `Transformer` ? Or am I missing something ?

Reply:
Oh, right, I thought this was one of your scala replacement classes.

What I have been doing for cases like this is throwing an `UnsupportedOperationException` in the body. It's not as good as not having the method, but it def. ensures it can't be used. And you don't have to maintain the code that's in the body.

Reply:
I understand that you cannot change the Java `Transformer` interface and must implement the deprecated method when calling `new Transformer` -- what I was wondering is about `scala.Transformer` interface -- should we add one and remove `punctuate` from it?

Reply:
FWIW, I think adding a new scala interface just to remove a method that we plan to remove from the java interface is not necessary. Better just to implement it and move on.

Also, it would be a bit trickier to swap in a scala replacement for `Transformer` than for the top-level DSL classes, since implementations of the java `Transformer` won't implement the scala `Transfomer`, so you wouldn't be able to plug them in via the scala DSL wrapper. But there's otherwise no reason this shouldn't work.

Reply:
Ack. Was just an idea. I don't really speak Scala (yet) -- this is an exercise to learn something about it...

If we need to have it, I vote to throw an exception to forces users to use the new API.